### DISCUSSION TITLE
SSL validation with ktor as http client

### DISCUSSION THREAD

**Discussion by N7ghtm4r3** [coil-kt/coil#2746] [Q&A] [answered]
Hi, I would use `ktor`as http client for an `KMP` application and I need to validate a Self Signed Certificates. This implementation is similar to the `OkHttp` client like here #2218? Thank you!

**Comment by colinrtwhite** [comment]
You can pass a custom `HttpClient` to `KtorNetworkFetcherFactory`. I'm not too sure about the config required to validate self signed certs.

**Reply by N7ghtm4r3** [reply]
Thank you! I will test this suggestion and I'll let you know

**Comment by N7ghtm4r3** [comment]
Hi, can you share an example snippet of the code? 
I struggling to integrate the `KtorNetworkFetcherFactory`, my code is the following:

```kotlin
imageLoader = ImageLoader.Builder(LocalPlatformContext.current)
        .components {
            addFetcherFactories {
                listOf(
                    KtorNetworkFetcherFactory()
                )
            }
        }
        .build() 
```

But there is a type mismatch:

```kotlin
Required:
List<Pair<coil3.fetch.Fetcher.Factory<out Any>, KClass<out Any>>>
Found:
List<coil3.network.NetworkFetcher.Factory>
```

**Reply by N7ghtm4r3** [reply]
Updated snippet:

```kotlin
imageLoader = ImageLoader.Builder(LocalPlatformContext.current)
        .components {
            addFetcherFactories {
                listOf(
                    Pair(
                        first = KtorNetworkFetcherFactory(
                            httpClient = customHttpClient()
                        ),
                        second = NetworkFetcher::class
                    )
                )
            }
        }
        .build()
``` 
### Android

```kotlin
actual fun customHttpClient(): HttpClient {
    val sslContext = SSLContext.getInstance("TLS")
    val certificates = validateSelfSignedCertificate()
    sslContext.init(null, certificates, SecureRandom())
    return HttpClient(OkHttp) {
        engine {
            config {
                sslSocketFactory(sslContext.socketFactory, certificates[0] as X509TrustManager)
                hostnameVerifier { _, _ -> true }
            }
        }
    }
}

private fun validateSelfSignedCertificate(): Array<TrustManager> {
    return arrayOf(@SuppressLint("CustomX509TrustManager")
    object : X509TrustManager {
        override fun getAcceptedIssuers(): Array<X509Certificate> {
            return arrayOf()
        }

        @SuppressLint("TrustAllX509TrustManager")
        override fun checkClientTrusted(certs: Array<X509Certificate>, authType: String) {}
        @SuppressLint("TrustAllX509TrustManager")
        override fun checkServerTrusted(certs: Array<X509Certificate>, authType: String) {}
    })
}
```

### Desktop

```kotlin
actual fun customHttpClient(): HttpClient {
    val sslContext = SSLContext.getInstance("TLS")
    sslContext.init(null, validateSelfSignedCertificate(), SecureRandom())
    return HttpClient(Java) {
        engine {
            config {
                sslContext(sslContext)
            }
        }
    }
}

private fun validateSelfSignedCertificate(): Array<TrustManager> {
    return arrayOf(@Suppress("CustomX509TrustManager")
    object : X509TrustManager {
        override fun getAcceptedIssuers(): Array<X509Certificate> {
            return arrayOf()
        }

        @Suppress("TrustAllX509TrustManager")
        override fun checkClientTrusted(certs: Array<X509Certificate>, authType: String) {}
        @Suppress("TrustAllX509TrustManager")
        override fun checkServerTrusted(certs: Array<X509Certificate>, authType: String) {}
    })
}
```

but still struggling to bypass the self-signed certs with the error: 
`ErrorResult(image=null, request=ImageRequest(context=coil3.PlatformContext$Companion$INSTANCE$1@28c32a6d, data=self-signed-url target=coil3.compose.AsyncImagePainter$updateRequest$$inlined$target$default$1@2d80903d, listener=null, memoryCacheKey=null, memoryCacheKeyExtras={}, diskCacheKey=null, fileSystem=NioSystemFileSystem, fetcherFactory=null, decoderFactory=null, interceptorCoroutineContext=EmptyCoroutineContext, fetcherCoroutineContext=Dispatchers.IO, decoderCoroutineContext=Dispatchers.IO, memoryCachePolicy=ENABLED, diskCachePolicy=ENABLED, networkCachePolicy=ENABLED, placeholderMemoryCacheKey=null, placeholderFactory=coil3.util.UtilsKt$EMPTY_IMAGE_FACTORY$1@56753a1b, errorFactory=coil3.util.UtilsKt$EMPTY_IMAGE_FACTORY$1@56753a1b, fallbackFactory=coil3.util.UtilsKt$EMPTY_IMAGE_FACTORY$1@56753a1b, sizeResolver=coil3.compose.ConstraintsSizeResolver@1c4bec82, scale=FILL, precision=INEXACT, extras=Extras(data={coil3.Extras$Key@73cab510=500}), defined=Defined(fileSystem=null, interceptorCoroutineContext=null, fetcherCoroutineContext=null, decoderCoroutineContext=null, memoryCachePolicy=null, diskCachePolicy=null, networkCachePolicy=null, placeholderFactory=coil3.util.UtilsKt$EMPTY_IMAGE_FACTORY$1@56753a1b, errorFactory=coil3.util.UtilsKt$EMPTY_IMAGE_FACTORY$1@56753a1b, fallbackFactory=coil3.util.UtilsKt$EMPTY_IMAGE_FACTORY$1@56753a1b, sizeResolver=coil3.compose.ConstraintsSizeResolver@1c4bec82, scale=FILL, precision=INEXACT), defaults=Defaults(fileSystem=NioSystemFileSystem, interceptorCoroutineContext=EmptyCoroutineContext, fetcherCoroutineContext=Dispatchers.IO, decoderCoroutineContext=Dispatchers.IO, memoryCachePolicy=ENABLED, diskCachePolicy=ENABLED, networkCachePolicy=ENABLED, placeholderFactory=coil3.util.UtilsKt$EMPTY_IMAGE_FACTORY$1@56753a1b, errorFactory=coil3.util.UtilsKt$EMPTY_IMAGE_FACTORY$1@56753a1b, fallbackFactory=coil3.util.UtilsKt$EMPTY_IMAGE_FACTORY$1@56753a1b, sizeResolver=RealSizeResolver(size=Size(width=Undefined, height=Undefined)), scale=FIT, precision=EXACT, extras=Extras(data={coil3.Extras$Key@7e24dabe=true}))), throwable=javax.net.ssl.SSLHandshakeException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target)`

**Reply by colinrtwhite** [reply, marked as answer]
The fetcher factory isn't being added correctly; [check out the docs](https://coil-kt.github.io/coil/network/#using-a-custom-okhttpclient).

**Reply by N7ghtm4r3** [reply]
I read the docs and updated the `imageloader `code snippet as the following:

```kotlin
imageLoader = ImageLoader.Builder(LocalPlatformContext.current)
        .components {
            add(
                KtorNetworkFetcherFactory(
                    httpClient = customHttpClient()
                )
            )
        }
        .build()
```
but an error still occurs 
`ErrorResult(image=null, request=ImageRequest(context=coil3.PlatformContext$Companion$INSTANCE$1@6ab14cce, data=https://217.160.205.94:1809/profiles/d91ae52f00b6468b8ea4d6ec77cb73571719665855244.jpg, target=coil3.compose.AsyncImagePainter$updateRequest$$inlined$target$default$1@6418d34c, listener=null, memoryCacheKey=null, memoryCacheKeyExtras={}, diskCacheKey=null, fileSystem=NioSystemFileSystem, fetcherFactory=null, decoderFactory=null, interceptorCoroutineContext=EmptyCoroutineContext, fetcherCoroutineContext=Dispatchers.IO, decoderCoroutineContext=Dispatchers.IO, memoryCachePolicy=ENABLED, diskCachePolicy=ENABLED, networkCachePolicy=ENABLED, placeholderMemoryCacheKey=null, placeholderFactory=coil3.util.UtilsKt$EMPTY_IMAGE_FACTORY$1@5af990b4, errorFactory=coil3.util.UtilsKt$EMPTY_IMAGE_FACTORY$1@5af990b4, fallbackFactory=coil3.util.UtilsKt$EMPTY_IMAGE_FACTORY$1@5af990b4, sizeResolver=coil3.compose.ConstraintsSizeResolver@6bbccc16, scale=FILL, precision=INEXACT, extras=Extras(data={coil3.Extras$Key@25290eab=500}), defined=Defined(fileSystem=null, interceptorCoroutineContext=null, fetcherCoroutineContext=null, decoderCoroutineContext=null, memoryCachePolicy=null, diskCachePolicy=null, networkCachePolicy=null, placeholderFactory=coil3.util.UtilsKt$EMPTY_IMAGE_FACTORY$1@5af990b4, errorFactory=coil3.util.UtilsKt$EMPTY_IMAGE_FACTORY$1@5af990b4, fallbackFactory=coil3.util.UtilsKt$EMPTY_IMAGE_FACTORY$1@5af990b4, sizeResolver=coil3.compose.ConstraintsSizeResolver@6bbccc16, scale=FILL, precision=INEXACT), defaults=Defaults(fileSystem=NioSystemFileSystem, interceptorCoroutineContext=EmptyCoroutineContext, fetcherCoroutineContext=Dispatchers.IO, decoderCoroutineContext=Dispatchers.IO, memoryCachePolicy=ENABLED, diskCachePolicy=ENABLED, networkCachePolicy=ENABLED, placeholderFactory=coil3.util.UtilsKt$EMPTY_IMAGE_FACTORY$1@5af990b4, errorFactory=coil3.util.UtilsKt$EMPTY_IMAGE_FACTORY$1@5af990b4, fallbackFactory=coil3.util.UtilsKt$EMPTY_IMAGE_FACTORY$1@5af990b4, sizeResolver=RealSizeResolver(size=Size(width=Undefined, height=Undefined)), scale=FIT, precision=EXACT, extras=Extras(data={coil3.Extras$Key@3b03b7b=true}))), throwable=javax.net.ssl.SSLHandshakeException: No subject alternative names present)`

From: `unable to find valid certification path to requested target`
To: `No subject alternative names present`

**Reply by N7ghtm4r3** [reply]
On `Android` works, so I need to refactor the `JVM` snippet. Thank you for the help!

**Reply by N7ghtm4r3** [reply]
Just for info I migrated from `implementation("io.ktor:ktor-client-java:3.0.2")` to `implementation("io.ktor:ktor-client-okhttp:3.0.2")` and mantained the same snippet of `Android` and it works fine! Thank you again
